fs-storage/src/folder_storage.rs

```diff
@@ -0,0 +1,784 @@
+    ///
+    /// Note: if the folder storage already exists, the data will be read from the folder
+    /// without overwriting it.
+    pub fn new(label: String, path: &Path) -> Result<Self> {
+        let mut storage = Self {
+            label,
+            path: PathBuf::from(path),
+            modified: BTreeMap::new(),
+            written_to_disk: BTreeMap::new(),
+            data: BTreeMap::new(),
+            deleted_keys: BTreeSet::new(),
+        };
+
+        if Path::exists(path) {
+            storage.read_fs()?;
+        }
+
+        Ok(storage)
+    }
+
+    /// Load mapping from folder storage
+    fn load_fs_data(&mut self) -> Result<()> {
+        if !self.path.exists() {
+            return Err(ArklibError::Storage(
+                self.label.clone(),
+                "Folder does not exist".to_owned(),
+            ));
+        }
+
+        if !self.path.is_dir() {
+            return Err(ArklibError::Storage(
+                self.label.clone(),
+                "Path is not a directory".to_owned(),
+            ));
+        }
+
+        let mut data = BTreeMap::new();
+
+        for entry in fs::read_dir(&self.path)? {
+            let entry = entry?;
+            let path = entry.path();
+            if path.is_file()
+                && path
+                    .extension()
+                    .map_or(false, |ext| ext == "json")
+            {
+                let key: K = self.extract_key_from_file_path(&path)?;
+                let file = File::open(&path)?;
+                let value: V =
+                    serde_json::from_reader(file).map_err(|err| {
+                        ArklibError::Storage(
+                            self.label.clone(),
+                            err.to_string(),
+                        )
+                    })?;
+
+                data.insert(key.clone(), value);
+
+                if let Ok(metadata) = fs::metadata(&path) {
+                    if let Ok(modified) = metadata.modified() {
+                        self.written_to_disk.insert(key.clone(), modified);
+                        self.modified.insert(key, modified);
+                    }
+                }
+            }
+        }
+
+        self.data = data;
+        Ok(())
+    }
+    /// Resolve differences between memory and disk data
+    fn resolve_divergence(&mut self) -> Result<()> {
+        let new_data = FolderStorage::new("new_data".into(), &self.path)?;
+
+        for (key, new_value) in new_data.data.iter() {
+            if let Some(existing_value) = self.data.get(key) {
+                let existing_value_updated = self
+                    .modified
+                    .get(key)
+                    .and_then(|ram_stamp| {
+                        self.written_to_disk
+                            .get(key)
+                            .map(|disk_stamp| ram_stamp > disk_stamp)
+                    })
+                    .unwrap_or(false);
```

Comment:
Why are we defaulting to `false` here?

Reply:
@twitu 

btw, defaulting to `true` also works

Reply:
The `existing_value_updated` triggers a monoidal combine of values for a particular key.

Right now it works because the property based test uses a Dummy value which has a "dumb" monoidal implementation which does nothing i.e. combine(dummy, dummy) = dummy.

However, when we use values like strings or numbers that have useful monoidal implementations, we want to appropriately set the `existing_value_updated` flag. We want to combine things only if values have diverged in both memory and storage. So the default should be false, if the value is not updated it is simply set and not combined.